clusterrmdup update

pontushojer opened the pull request:
We had some issues with running `clusterrmdup` for large dataset (see #173 ) so I have done some updates to limit memory and increase performance. I also changed some aspects that would cause the script to miss some duplicates based on the [picard definition for duplicates](https://sourceforge.net/p/samtools/mailman/message/25062576/): 

"... find the 5' coordinates and mapping orientations of each read pair.  [...].  It then matches all read pairs that have identical 5' coordinates and orientations and marks as duplicates all but the "best" pair."

The previous script used the 5' and 3' positions of both reads and did not consider the orientation. Thus I have now changed so that only 5' positions and orientation is considered. The image below shows one such scenario were a duplicate between barcodes GTCCTTTGTTCCGTACTAAG and AATAGGTCTTTCGACATACG but the 3' end of read2 is trimmed differently. "test.master.bam" this previous script which miss this duplicate while the new script ("test.new.bam") detects it and merges the barcodes.    

![image](https://user-images.githubusercontent.com/27061883/70517194-7b5c8180-1b38-11ea-871a-8651d4d43992.png)



I have also had to implement a buffer system to be able to catch all read on a duplicate position. Consider the following scenario.
```
READ PAIR 1:    R1|--------->    <-------|R2
READ PAIR 2:    R1|--------->       <----|R2
Chromosome:  ==================================>
```

Buy the definition stated above the two read pairs (RP) are duplicates but R2 for RP2 is on a different position start position compared to R2 in RP1. Thus it will appear longer down in the coordinate sorted BAM file. The previous script looked at each position separately and would thus drop RP1 as possible duplicate. Now instead each RP within a defined window (currently 500 bp) is considered. Thus these cases can be caught and handled correctly.

**Other changes**
- Move `summary` away from global scope (see #167 )
- Factored out several functions.
- Name changes  

**Runtime for chr22 testdata.** 
Previous script
- real: 10m18.406s, 10m21.591s
- user: 10m16.232s, 10m19.550s
- sys: 0m1.316s, 0m1.035s

This script
- real: 8m3.786s, 8m6.691s
- user: 8m1.998s, 8m5.079s
- sys: 0m1.052s, 0m0.900s

FrickTobias commented on src/blr/cli/clusterrmdup.py:
```diff
@@ -12,237 +12,297 @@
+        if mate.is_read1 and read.is_read2:
+            orientation = "F"
+        else:
+            orientation = "R"
```
When I read this I first confused this `orientation` with `orientation_ok`, maybe one could specify this as `rp_orientation`?

pontushojer replied:
I changed the name of `orientation_ok` instead to `valid_pair_orientation`.

FrickTobias commented on src/blr/cli/clusterrmdup.py:
```diff
@@ -12,237 +12,297 @@
-                mate = cache_reads[read.query_name]
-                del cache_reads[read.query_name]
+    buffer_dup_pos = deque()
+    for read, mate in tqdm(parse_filtered_read_pairs(args.input, summary), desc="Reading pairs"):
```
I think this makes it seem as the function expects a filtered input, how about renaming it as `parse_and_filter_read_pairs`?

pontushojer replied:
True

FrickTobias commented on src/blr/cli/clusterrmdup.py:
```diff
@@ -12,237 +12,297 @@
+        barcode = utils.get_bamtag(read, args.barcode_tag)
+        if not barcode:
+            summary["Non tagged reads"] += 2
+            continue
```
This should be moved to `meet_requirements()` now when the reader filters reads.

FrickTobias commented on src/blr/cli/clusterrmdup.py:
```diff
@@ -12,237 +12,297 @@
-                        current_cache_rp[rp_pos_tuple] = CacheReadPairTracker(rp_pos_tuple=rp_pos_tuple,
-                                                                              chromosome=chrom_new, read=read,
-                                                                              mate=mate, bc=bc_new)
+        if abs(pos_new - pos_prev) > BUFFER_SIZE or not chrom_new == chrom_prev:
```
```suggestion
        if abs(pos_new - pos_prev) > BUFFER_SIZE or chrom_new  not == chrom_prev:
```

FrickTobias replied:
Minor comment: Do we need such a big buffer size? It shouldn't be able to be more than a couple of bp right?

FrickTobias commented on src/blr/cli/clusterrmdup.py:
```diff
@@ -12,237 +12,297 @@
+        self.reads = int()
         self.barcodes = set()
+        self.checked = False
+        self.updated = bool()
```
```suggestion
        self.contains_reads = bool()
```

pontushojer replied:
I think `updated` makes more sense here although it may not be obvious. As the position is only removed when its outside the buffer the same position may be evaluated multiple times. I added this so that once it had been evaluated it would only be evaluated again if more reads had been added. Thus is called `updated`.

FrickTobias replied:
Ah, I see. Makes sense!

pontushojer commented:
Related to a question I compared how the changing the buffer size affects the script. The current default is 500 bp.

Findings:
- Runtime unaffected by buffer size in 5-50,000 bp range. 
- Low buffer size --> Low memory but less sensitive
- High buffer size --> Increased memory but more sensitive

Sensitivity in this case relates to the ability to detect and remove duplicate barcode clusters (see "Barcodes removed" in image). Thus there is some sort of tradeoff between memory and sensitivity, therefore I think 500 it still reasonable. 

![image](https://user-images.githubusercontent.com/27061883/70537367-9c38cd00-1b60-11ea-8374-0151695da52b.png)

FrickTobias commented:
Are you sure the plot is correct? The barcodes removed are at a maximum when `BUFFER_SIZE` is the lowest, suggesting sensitivity is high (assuming they are TP).

pontushojer commented:
> Are you sure the plot is correct? The barcodes removed are at a maximum when `BUFFER_SIZE` is the lowest, suggesting sensitivity is high (assuming they are TP).

Some more investigation into this showed a small error in the code that caused duplicate positions to be checked against themself. This was increasingly prominent for low buffer sizes where `find_barcode_duplicates` is called more often. The error is now fixed and I can show the current data

![image](https://user-images.githubusercontent.com/27061883/70618508-54bf4900-1c13-11ea-961b-92e3e079c16d.png)

Here we instead see an increase in sensitivity with increased buffer size at the expense of increased memory use. Based on this I changed the default buffer size to 200 bp which is close  just above the typical read length (150 bp). I also added buffer_size as an optional argument.

FrickTobias commented:
Awesome! Looks good!
